api/v1alpha1/catalogentry_types.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	kcpv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// InternalAPIBindingEntryLabelKey is the label key on an APIBinding with the
+	// base62(sha224(<workspace>:<exportName>)) as value to filter bindings by catalog entry.
+	InternalAPIBindingEntryLabelKey = "internal.catalog.kcp.dev/entry"
+)
+
+// These are valid conditions of CatalogEntry.
+const (
+	// CatalogEntryValid is a condition for CatalogEntry that reflects the validity
+	// of the referenced APIExport.
+	CatalogEntryValid conditionsv1alpha1.ConditionType = "CatalogEntryValid"
+	// CatalogEntryInvalidReferenceReason is a reason for the CatalogEntryValid
+	// condition of APIBinding that the referenced CatalogEntry reference is invalid.
+	CatalogEntryInvalidReferenceReason = "CatalogEntryInvalidReference"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// CatalogEntry is the Schema for the catalogentries API
+type CatalogEntry struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   CatalogEntrySpec   `json:"spec,omitempty"`
+	Status CatalogEntryStatus `json:"status,omitempty"`
+}
+
+// CatalogEntrySpec defines the desired state of CatalogEntry
+type CatalogEntrySpec struct {
+	// References uniquely identify a list of APIs to bind to
+	References []kcpv1alpha1.ExportReference `json:"references"`
+	// maximalPermissionPolicy will allow for a service provider to set an upper bound on what is allowed
+	// for a consumer of this API. If the policy is not set, no upper bound is applied,
+	// i.e the consuming users can do whatever the user workspace allows the user to do.
+	//
+	// The policy consists of RBAC (Cluster)Roles and (Cluster)Bindings. A request of a user in
+	// a workspace that binds to this APIExport via an APIBinding is additionally checked against
+	// these rules, with the user name and the groups prefixed with `apis.kcp.dev:binding:`.
+	//
+	// +optional
+	MaximalPermissionPolicy *kcpv1alpha1.MaximalPermissionPolicy `json:"maximalPermissionPolicy,omitempty"`
```

Comment:
Is this copied from the APIExport references? What is it for / how is it used?

Reply:
This is a copy-over from APIExport as it is a suggestion from Stefan on having a maximum cap on the permissions that the entry may have. If the permissions from APIExport exceeds the max settings here, it should mark the entry as invalid and present it from being to be able to bind. This max settings can be used as a way to safeguard the potential changes to ExportReference list which may exceed the appropriate permissions. This should conform with minimal permission principle.